File: TelerikUIExamples/TelerikUIExamples/ListViewAnimations.m
//
//  ListViewAnimations.m
//  TelerikUIExamples
//
//  Copyright (c) 2015 Telerik. All rights reserved.
//

#import <TelerikUI/TelerikUI.h>
#import "ListViewAnimations.h"
#import "AnimationListCell.h"

@interface ListViewAnimations ()
{
    TKListView *_listView;
    TKDataSource *_dataSource;
    NSMutableArray *_images;
}

@end

@implementation ListViewAnimations

- (instancetype)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil
{
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self) {
        [self addOption:@"Scale in" selector:@selector(scaleInSelected)];
        [self addOption:@"Fade in" selector:@selector(fadeInSelected)];
        [self addOption:@"Slide in" selector:@selector(slideInSelected)];
    }
    return self;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    _dataSource = [TKDataSource new];
    [_dataSource loadDataFromJSONResource:@"ListViewSampleData" ofType:@"json" rootItemKeyPath:@"photos"];
    
    _images = [NSMutableArray new];
    for (int i = 0; i<_dataSource.items.count; i++) {
        [_images addObject:[UIImage imageNamed:_dataSource.items[i]]];
    }
    
    [_dataSource.settings.listView createCell:^TKListViewCell *(TKListView *listView, NSIndexPath *indexPath, id item) {
        return [listView dequeueReusableCellWithReuseIdentifier:@"cell" forIndexPath:indexPath];
    }];
    
    [_dataSource.settings.listView initCell:^(TKListView *listView, NSIndexPath *indexPath, TKListViewCell *cell, id item) {
        cell.imageView.image = _images[indexPath.row];
    }];
    
    _listView = [[TKListView alloc] initWithFrame:self.view.bounds];
    _listView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
    _listView.dataSource = _dataSource;
    [_listView registerClass:[AnimationListCell class] forCellWithReuseIdentifier:@"cell"];
    [self.view addSubview:_listView];

    TKListViewGridLayout *layout = [TKListViewGridLayout new];
    if ([UIDevice currentDevice].userInterfaceIdiom == UIUserInterfaceIdiomPad) {
        layout.spanCount = 3;
    }
    else {
        layout.spanCount = 2;
    }
    layout.itemSize = CGSizeMake(130, 180);
    layout.itemSpacing = 10;
    layout.lineSpacing = 10;
    
    // >> listview-alignment
    layout.itemAlignment = TKListViewItemAlignmentCenter;
    // << listview-alignment
    
    layout.itemAppearAnimation = TKListViewItemAnimationScale;
    
    // >> listview-animation-duration
    layout.animationDuration = 0.4;
    // << listview-animation-duration
    
    _listView.layout = layout;
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)fadeInSelected
{
    TKListViewLinearLayout *layout = (TKListViewLinearLayout*)_listView.layout;
    layout.itemAppearAnimation = TKListViewItemAnimationFade;
    layout.itemInsertAnimation = TKListViewItemAnimationFade;
    layout.itemInsertAnimation = TKListViewItemAnimationFade;
}

- (void)slideInSelected
{
    TKListViewLinearLayout *layout = (TKListViewLinearLayout*)_listView.layout;
    layout.itemAppearAnimation = TKListViewItemAnimationSlide;
    layout.itemInsertAnimation = TKListViewItemAnimationSlide;
    layout.itemDeleteAnimation = TKListViewItemAnimationSlide;
}

- (void)scaleInSelected
{
    TKListViewLinearLayout *layout = (TKListViewLinearLayout*)_listView.layout;
    // >> listview-appear
    layout.itemAppearAnimation = TKListViewItemAnimationScale;
    // << listview-appear
    
    // >> listview-insert
    layout.itemInsertAnimation = TKListViewItemAnimationScale;
    // << listview-insert
    
    // >> listview-delete
    layout.itemDeleteAnimation = TKListViewItemAnimationScale;
    // << listview-delete
    
}

@end

File: Examples/Examples/ListView/ListViewAnimations.cs
using System;

using Foundation;
using UIKit;
using CoreGraphics;

using TelerikUI;

namespace Examples
{
	[Register("ListViewAnimations")]
	public class ListViewAnimations: XamarinExampleViewController
	{
		TKDataSource dataSource = new TKDataSource();
		TKListView listView = new TKListView();

		public override void ViewDidLoad ()
		{
			AddOption ("Scale in", ScaleInSelected);
			AddOption ("Fade in", FadeInSelected);
			AddOption ("Slide in", SlideInSelected);

			base.ViewDidLoad ();

			string s = TKChart.VersionString;

			this.dataSource.LoadDataFromJSONResource ("ListViewSampleData", "json", "photos");

			this.dataSource.Settings.ListView.CreateCell ((TKListView listView, NSIndexPath indexPath, NSObject item) => {
				return listView.DequeueReusableCell("cell", indexPath) as TKListViewCell;
			});
			this.dataSource.Settings.ListView.InitCell (delegate (TKListView listView, NSIndexPath indexPath, TKListViewCell cell, NSObject item) {
				cell.ImageView.Image = UIImage.FromBundle (this.dataSource.Items[indexPath.Row] as NSString);
			});

			this.listView.Frame = this.View.Bounds;
			this.listView.AutoresizingMask = UIViewAutoresizing.FlexibleWidth | UIViewAutoresizing.FlexibleHeight;
			this.listView.WeakDataSource = this.dataSource;
			this.listView.RegisterClassForCell (new ObjCRuntime.Class (typeof(AnimationListCell)), "cell");
			this.View.AddSubview (this.listView);

			var layout = new TKListViewGridLayout();
			if (UIDevice.CurrentDevice.UserInterfaceIdiom == UIUserInterfaceIdiom.Pad) {
				layout.SpanCount = 3;
			}
			else {
				layout.SpanCount = 2;
			}
			layout.ItemSize = new CGSize(130, 180);
			layout.ItemSpacing = 10;
			layout.LineSpacing = 10;

			// >> listview-alignment-cs
			layout.ItemAlignment = TKListViewItemAlignment.Center;
			// << listview-alignment-cs

			layout.ItemAppearAnimation = TKListViewItemAnimation.Scale;

			// >> listview-animation-duration-cs
			layout.AnimationDuration = 0.4f;
			// << listview-animation-duration-cs

			this.listView.Layout = layout;
		}

		void ScaleInSelected()
		{
			TKListViewLinearLayout layout = (TKListViewLinearLayout)this.listView.Layout;
			// >> listview-appear-cs
			layout.ItemAppearAnimation = TKListViewItemAnimation.Scale;
			// << listview-appear-cs

			// >> listview-insert-cs
			layout.ItemInsertAnimation = TKListViewItemAnimation.Scale;
			// << listview-insert-cs

			// >> listview-delete-cs
			layout.ItemDeleteAnimation = TKListViewItemAnimation.Slide;
			// << listview-delete-cs
		}

		void FadeInSelected()
		{
			TKListViewLinearLayout layout = (TKListViewLinearLayout)this.listView.Layout;
			layout.ItemAppearAnimation = TKListViewItemAnimation.Fade;
			layout.ItemInsertAnimation = TKListViewItemAnimation.Fade;
			layout.ItemDeleteAnimation = TKListViewItemAnimation.Fade;
		}

		void SlideInSelected()
		{
			TKListViewLinearLayout layout = (TKListViewLinearLayout)this.listView.Layout;
			layout.ItemAppearAnimation = TKListViewItemAnimation.Slide;
			layout.ItemInsertAnimation = TKListViewItemAnimation.Slide;
			layout.ItemDeleteAnimation = TKListViewItemAnimation.Slide;
		}
	}
}

